### PR TITLE
C2LC-173: Implement an action panel component

### PR DESCRIPTION
**Pull request by chosww**

### CODE REVIEW

**Comment by simonbates** [comment]
I merged the develop-0.3 branch (after merging the C2LC-180 PR) into C2LC-173 and did some refactoring on the ActionPanel component:

https://github.com/simonbates/c2lc-coding-environment/commit/55d321671856522677583a520f7761c8c2dd5b21

Please merge my CL2C-173 into your branch to update this PR.

**Comment by simonbates** [comment]
I added some tests to check boundary conditions to `ActionPanel.test.js`: move backwards when at the start of the program and move forwards when at the end. In the designs, the action panel movement buttons should be disabled in those cases.

In the current implementation, the buttons are enabled but do nothing. Please update the `ActionPanel` to use `AriaDisablingButton` and disable the movement buttons in these cases.

The tests that I added to `ActionPanel.test.js` are passing and reflect the current behaviour, rather than the behaviour outlined above. I left `TODO` comments with the desired behaviour:

- https://github.com/simonbates/c2lc-coding-environment/blob/55d321671856522677583a520f7761c8c2dd5b21/src/ActionPanel.test.js#L95-L96
- https://github.com/simonbates/c2lc-coding-environment/blob/55d321671856522677583a520f7761c8c2dd5b21/src/ActionPanel.test.js#L117-L118

Update the tests as appropriate.

**Comment by simonbates** [comment]
I found a bug when inserting into a program with the action panel open. To reproduce:

1. Make a program with a single move forward step
1. Open the action panel for the step
1. Drag a turn left tile to before the move forward tile

Expected:

- The action panel stays open and stays connected to the move forward tile

Actual:

- The action panel stays open, but is now connected to the newly inserted tile

**Comment by simonbates** [comment]
Apologies. Please ignore the `TODO` comments to special case the ARIA labels:

- https://github.com/simonbates/c2lc-coding-environment/blob/55d321671856522677583a520f7761c8c2dd5b21/src/ActionPanel.test.js#L96
- https://github.com/simonbates/c2lc-coding-environment/blob/55d321671856522677583a520f7761c8c2dd5b21/src/ActionPanel.test.js#L118

I had missed that the designs specify the ARIA labels and that what you have implemented matches the designs.

**Comment by simonbates** [comment]
On Friday June 26, Daniel, Sepideh, and Simon discussed changes to how and when the new control panel is closed. These notes are captured in a Google doc:

https://docs.google.com/document/d/1r5u2id18HDTucbhFzcT5KN4H3emG9CV-OvI2APt4YDs/edit

For 0.3, we will implement Option 5: Close control panel when interacting with something other than the control panel and the movements palette, and when we initiate a drag operation.

This change will address the bug described above: https://github.com/codelearncreate/c2lc-coding-environment/pull/51#issuecomment-649757256

https://issues.fluidproject.org/browse/C2LC-173?focusedCommentId=41721&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-41721

**Comment by chosww** [comment]
For the work to change the action panel callbacks to use index input, there's a few ways to bind the callbacks to ActionPanel component, and my implementation uses arrow function. 
Options can be found at: https://reactjs.org/docs/faq-functions.html

File: src/ActionPanel.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                id='ActionPanel'
+                className={'ActionPanel__panel'}
+                data-actionpanelgroup={true}
+                ref={this.actionPanelRef}>
+                <AriaDisablingButton
+                    name='deleteCurrentStep'
+                    disabled={false}
+                    aria-label={this.props.intl.formatMessage({id:'ActionPanel.action.delete'}, stepInfoMessage)}
+                    className='ActionPanel__action-buttons'
+                    onClick={()=>{this.props.onDelete(this.props.pressedStepIndex)}}>
```
Comment (simonbates) [inline src/ActionPanel.js:86]: Change the `onClick` handlers for the action panel buttons to be consistent with our other components:

    onClick={this.handleClickDelete}

And implement `handle...` click handlers for each of the buttons.

File: src/ActionPanel.test.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+describe('ActionPanel options', () => {
+    test('When the deleteCurrentStep option is selected on second step turn left of the program', () => {
+        const { wrapper, mockDeleteHandler } = createMountActionPanel();
+        const deleteButton = getActionPanelOptionButtons(wrapper, 'deleteCurrentStep');
+        const expectedAriaLabel = 'Delete Step 2 turn left';
+        expect(deleteButton.get(0).props['aria-label']).toBe(expectedAriaLabel);
+        deleteButton.simulate('click');
+        expect(mockDeleteHandler.mock.calls.length).toBe(1);
```
Comment (simonbates) [inline src/ActionPanel.test.js:69]: Check that the callback is called with the expected program step index.

File: src/ActionPanel.test.js [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    });
+
+    test('When the replaceCurrentStep option is selected on second step turn left of the program', () => {
+        const { wrapper, mockReplaceHandler } = createMountActionPanel();
+        const replaceCurrentStepButton = getActionPanelOptionButtons(wrapper, 'replaceCurrentStep');
+        const expectedAriaLabel = 'Replace Step 2 turn left with selected action right';
+        expect(replaceCurrentStepButton.get(0).props['aria-label']).toBe(expectedAriaLabel);
+        replaceCurrentStepButton.simulate('click');
+        expect(mockReplaceHandler.mock.calls.length).toBe(1);
```
Comment (simonbates) [inline src/ActionPanel.test.js:78]: Check that the callback is called with the expected program step index.

File: src/ActionPanel.test.js [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    })
+
+    test('When the moveToPreviousStep option is selected on second step turn left of the program', () => {
+        const { wrapper, mockMoveToPreviousStep } = createMountActionPanel();
+        const moveToPreviousStepButton = getActionPanelOptionButtons(wrapper, 'moveToPreviousStep');
+        const expectedAriaLabel = 'Move Step 2 turn left before step 1 forward';
+        expect(moveToPreviousStepButton.get(0).props['aria-label']).toBe(expectedAriaLabel);
+        moveToPreviousStepButton.simulate('click');
+        expect(mockMoveToPreviousStep.mock.calls.length).toBe(1);
```
Comment (simonbates) [inline src/ActionPanel.test.js:87]: Check that the callback is called with the expected program step index.

File: src/ActionPanel.test.js [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        expect(mockMoveToPreviousStep.mock.calls.length).toBe(0);
+    });
+
+    test('When the moveToNextStep option is selected on second step turn left of the program', () => {
+        const { wrapper, mockMoveToNextStep } = createMountActionPanel();
+        const moveToNextStepButton = getActionPanelOptionButtons(wrapper, 'moveToNextStep');
+        const expectedAriaLabel = 'Move Step 2 turn left after step 3 turn right';
+        expect(moveToNextStepButton.get(0).props['aria-label']).toBe(expectedAriaLabel);
+        moveToNextStepButton.simulate('click');
+        expect(mockMoveToNextStep.mock.calls.length).toBe(1);
```
Comment (simonbates) [inline src/ActionPanel.test.js:109]: Check that the callback is called with the expected program step index.
